Refactor: Replace subclassing FrameDiffer with subclassing Differ

wmanley opened the pull request:
Instead of subclassing `FrameDiffer` we now subclass `Differ` and pass an
instance of that subclass to `FrameDiffer`.  So before you'd write:

    press_and_wait.differ = BGRDiff

and now it's

    press_and_wait.differ = BGRDiff()

This means that we can customise differ parameters by passing arguments to
the constructor like:

    press_and_wait.differ = BGRDiff(threshold=60)

To support this differs are now immutable.  The `previous_frame` state that
used to live in `BGRDiff` and `GreyscaleDiff` is now implemented in
`FrameDiffer` which is now a concrete class.

We still need to support passing differ parameters directly to
`press_and_wait`, etc. like `press_and_wait(min_size=...)`, so there's a
`.replace` argument which returns a new differ, but with those parameters
set.

`Differ` is designed so we can efficiently implement `GrayscaleDiff`
without doing the colour -> grayscale conversion on the same image twice -
just as we could before.  To avoid internal state the interface makes
preprocessing explicit.  This makes the classes harder to use directly, but
that's ok because users won't be using them directly.

Currently we don't override `preprocess_mask` anywhere, but it seems to
fall out naturally from the design.

**TODO**

* [x] Rebase now that #785 is merged
* [x] Make tests pass
* ~~Pass ImageLogger into methods?~~

**Future work**

* Allow passing a `Differ` as `second_pass_method` to match.
* Global stbt.default_differ?
* Pass `differ` argument into `press_and_wait`, etc.

drothlis reviewed (commented):
Haven't finished the review, just jotting down some thoughts.

So if you want to implement a new diffing algorithm you subclass Differ and it needs to accept `min_size`, `threshold`, and `erode`. It can have new parameters too, but it must accept those.

It feels a bit weird to have `min_size` be part of each diff algorithm, but I suppose it's better to repeat that logic in each differ, than to repeat it in each caller (press_and_wait, etc). (Note: This isn't new, just an observation about the existing code.)

FrameDiffer should be removed from the public API, and Differ added to the public API (so that you can create your own subclasses — that's the only reason FrameDiffer was there).

There might be a better name for FrameDiffer. Its responsibilities:

- Remember the work done on seen frames.
- The logic for when we update the "reference" frame, which we added to
  help it detect slow gradual motion.

It's really the core of the detect_motion algorithm, but in an API that's more convenient to use from press_and_wait.

I'm not a fan of `preprocess_mask` — why do we need a wrapper around `load_mask` / `Mask.to_array`, do you have a concrete plan / use case for it?

Is it that some future differ might want a 3-channel mask? Both of the current differs want a single-channel mask, which is what the default implementation of `Differ.preprocess_mask` gives.

I don't like passing `mask_tuple` —I don't like passing ad-hoc tuples around, really— when we could pass it as `mask_pixels` and `region`.

Note that `Mask.to_array` is memoized, so we could call it from `BGRDiffer.{preprocess,diff}` directly? No, that feels wrong.

drothlis commented on _stbt/diff.py:
```diff
@@ -49,20 +49,77 @@ def __repr__(self) -> str:
+    preprocessing explicit.  This makes the classes hard to use, but that's ok -
+    it's a low-level interface intended mostly for internal use.
+
+    The only public interface is the subclass constructors.  The methods are not
+    intended for use by external code.
```
Are we going to support users implementing their own Differ subclasses? If so, they'll have to implement `preprocess` and `diff`, so those are technically public interfaces too.

drothlis replied:
Probably not — the interface is too complex. The real reason for making the base class public is for type annotations.

drothlis commented:
@wmanley I pushed a few commits implementing my earlier comments & tidying up the docstrings for public APIs. Let me know what you think.

wmanley commented:
I dropped your commit where `PreProcessedMask` was removed and moved the implementations of `preprocess_mask` into the concrete classes, hopefully to make it clearer.  I've also made the types "protected" with a leading underscore, hopefully to make the design clearer.  What do you think now?

drothlis reviewed (approved):
I'm a bit concerned about the complexity here: We have the `preprocess_mask` for the sake of mirroring `preprocess_frame`, but no specific use-case for it; a hypothetical way of typing them that we don't actually use (we aren't specifying the return types in the subclasses) and that doesn't type-check (and maybe isn't possible to actually express with Python's typing?); and a parameter called `mask` (in the subclasses) that isn't a `Mask`. I suppose it's not so bad, hopefully I'll still understand it next time I come to work in this area. :-)

drothlis commented:
On a more positive note, I'm excited about the possibility of passing custom differs into various APIs, including the `match` second pass!
